lib/async_hooks.js

```diff
@@ -255,18 +255,15 @@ class AsyncLocalStorage {
     resource[this.kResourceStore] = store;
   }
 
-  runSyncAndReturn(store, callback, ...args) {
-    const resource = executionAsyncResource();
-    const outerStore = resource[this.kResourceStore];
-    this.enterWith(store);
-    try {
+  run(store, callback, ...args) {
+    const resource = new AsyncResource('AsyncLocalStorage');
+    return resource.runInAsyncScope(() => {
+      this.enterWith(store);
       return callback(...args);
-    } finally {
-      resource[this.kResourceStore] = outerStore;
-    }
+    });
   }
 
-  exitSyncAndReturn(callback, ...args) {
+  exit(callback, ...args) {
     if (!this.enabled) {
```

Comment:
Should this also move to using `new AsyncResource()`?

Reply:
I don't think it needs one because the whole point _there_ is that it _doesn't_ propagate the context.